app/scripts/migrations/052.js

```diff
@@ -0,0 +1,163 @@
+import { cloneDeep, uniqBy } from 'lodash';
+import {
+  GOERLI,
+  GOERLI_CHAIN_ID,
+  KOVAN,
+  KOVAN_CHAIN_ID,
+  MAINNET,
+  MAINNET_CHAIN_ID,
+  NETWORK_TYPE_RPC,
+  RINKEBY,
+  RINKEBY_CHAIN_ID,
+  ROPSTEN,
+  ROPSTEN_CHAIN_ID,
+} from '../../../shared/constants/network';
+
+const version = 52;
+
+/**
+ * Migrate tokens in Preferences to be keyed by chainId instead of
+ * providerType. To prevent breaking user's MetaMask and selected
+ * tokens, this migration copies the RPC entry into *every* custom RPC
+ * chainId.
+ */
+export default {
+  version,
+  async migrate(originalVersionedData) {
+    const versionedData = cloneDeep(originalVersionedData);
+    versionedData.meta.version = version;
+    const state = versionedData.data;
+    versionedData.data = transformState(state);
+    return versionedData;
+  },
+};
+
+function transformState(state = {}) {
+  if (state.PreferencesController) {
+    const {
+      accountTokens,
+      accountHiddenTokens,
+      frequentRpcListDetail,
+    } = state.PreferencesController;
+
+    const newAccountTokens = {};
+    const newAccountHiddenTokens = {};
+
+    if (accountTokens && Object.keys(accountTokens).length > 0) {
+      for (const account of Object.keys(accountTokens)) {
```

Comment:
Nit: I think the account variable might be more appropriately name `address` or `accountAddress`, which better conveys what that variable is.

Reply:
done in the latest